src/libpgmoneta/wf_backup_incremental.c

```diff
@@ -0,0 +1,2018 @@
+   pgmoneta_free_message(basebackup_msg);
+   pgmoneta_free_message(tablespace_msg);
+   pgmoneta_free_query_response(response);
+   pgmoneta_token_bucket_destroy(bucket);
+   pgmoneta_token_bucket_destroy(network_bucket);
+   free(manifest_path);
+   free(chkptpos);
+   free(tag);
+   free(wal);
+
+   return 0;
+
+error:
+
+   if (backup_base == NULL)
+   {
+      backup_base = pgmoneta_get_server_backup_identifier(server, label);
+   }
+
+   if (pgmoneta_exists(backup_base))
+   {
+      pgmoneta_delete_directory(backup_base);
+   }
+
+   pgmoneta_close_ssl(ssl);
+   if (socket != -1)
+   {
+      pgmoneta_disconnect(socket);
+   }
+   pgmoneta_memory_destroy();
+   pgmoneta_memory_stream_buffer_free(buffer);
+   pgmoneta_free_message(tablespace_msg);
+   pgmoneta_free_tablespaces(tablespaces);
+   pgmoneta_free_message(basebackup_msg);
+   pgmoneta_free_query_response(response);
+   pgmoneta_token_bucket_destroy(bucket);
+   pgmoneta_token_bucket_destroy(network_bucket);
+   free(manifest_path);
+   free(chkptpos);
+   free(tag);
+   free(wal);
+
+   return 1;
+}
+
+static size_t
+get_incremental_header_size(uint32_t num_incr_blocks)
+{
+   size_t result;
+
+   /*
+       compute header size
+       (magic_number, truncation block length, block count) followed by block numbers
+    */
+   result = 3 * sizeof(uint32_t) + (sizeof(block_number) * num_incr_blocks);
+   /* round the header to a multiple of Block Size */
+   if ((num_incr_blocks > 0) && (result % block_size != 0))
+   {
+      result += block_size - (result % block_size);
+   }
+
+   return result;
+}
+
+static size_t
+get_incremental_file_size(uint32_t num_incr_blocks)
+{
+   size_t result;
+
+   result = get_incremental_header_size(num_incr_blocks);
+   result += block_size * num_incr_blocks;
+
+   return result;
+}
+
+static int
+compare_block_numbers(const void* a, const void* b)
+{
+   block_number aa = *(block_number*)a;
+   block_number bb = *(block_number*)b;
+
+   if (aa < bb)
+   {
+      return -1;
+   }
+   if (aa > bb)
+   {
+      return 1;
+   }
+   return 0;
+}
+
+static int
+create_standard_directories(int server, SSL* ssl, int socket, char* user, char* backup_data, char*** p, int* c)
+{
+   char** total_paths = NULL;
+   int total_count = 0;
+   char*** merged = NULL;
+   int* merged_count = NULL;
+   char* standard_dir = NULL;
+   int number_of_directories;
+   int number_of_files;
+   int valid_idx = 0;
+
+   number_of_directories = (sizeof(STANDARD_DIRECTORIES) / sizeof(char*));
+   number_of_files = (sizeof(STANDARD_FILES) / sizeof(char*));
+
+   merged = (char***)calloc((number_of_directories + 1 + 1), sizeof(char**));
+   merged_count = (int*)calloc((number_of_directories + 1 + 1), sizeof(int));
+
+   for (int i = 0; i < number_of_directories; i++)
+   {
+      char** paths = NULL;
+      int count = 0;
+
+      /* collect all the files inside this directory */
+      if (get_all_files(ssl, socket, STANDARD_DIRECTORIES[i], &paths, &count))
```

Comment:
Are we actually copying all the files here? How do we make sure incremental files are skipped?